parse.y

```diff
@@ -7564,35 +7631,20 @@ parser_has_token(struct parser_params *p)
     return pcur > ptok;
 }
 
-static VALUE
-code_loc_to_ary(struct parser_params *p, const rb_code_location_t *loc)
-{
-    VALUE ary = rb_ary_new_from_args(4,
-        INT2NUM(loc->beg_pos.lineno), INT2NUM(loc->beg_pos.column),
-        INT2NUM(loc->end_pos.lineno), INT2NUM(loc->end_pos.column));
-    rb_obj_freeze(ary);
-
-    return ary;
-}
-
 static void
-parser_append_tokens(struct parser_params *p, VALUE str, enum yytokentype t, int line)
-{
-    VALUE ary;
-    int token_id;
-
-    ary = rb_ary_new2(4);
-    token_id = p->token_id;
-    rb_ary_push(ary, INT2FIX(token_id));
-    rb_ary_push(ary, ID2SYM(parser_token2id(p, t)));
-    rb_ary_push(ary, str);
-    rb_ary_push(ary, code_loc_to_ary(p, p->yylloc));
-    rb_obj_freeze(ary);
-    rb_ary_push(p->tokens, ary);
+parser_append_tokens(struct parser_params *p, rb_parser_string_t *str, enum yytokentype t, int line)
+{
+    rb_parser_ast_token_t *token = xcalloc(1, sizeof(rb_parser_ast_token_t));
+    token->id = p->token_id;
+    token->type_id = parser_token2id(p, t);
+    token->str = str;
+    token->loc.beg_pos = p->yylloc->beg_pos;
+    token->loc.end_pos = p->yylloc->end_pos;
+    rb_parser_ary_push(p, p->tokens, (VALUE)token);
     p->token_id++;
 
     if (p->debug) {
-        rb_parser_printf(p, "Append tokens (line: %d) %"PRIsVALUE"\n", line, ary);
+        rb_parser_printf(p, "Append tokens (line: %d) %"PRIsVALUE"\n", line, (VALUE)token);
```

Comment:
`PRsVALUE` calls `to_s` method on the corresponding argument.
`rb_parser_ast_token_t` does not seem to compatible with `VALUE`.

Reply:
Thank you for the review!
I implemented `rb_parser_str_escape()` which does not depend on VALUE
https://github.com/ruby/ruby/pull/10004/commits/15e42bcc75f65a7ccd7dc21f569c8f567c31ee35

Reply:
@nobu Could you please review them again?